Fix part of #2869: Add mobile feedback button to exploration player.

Oishikatta opened the pull request:
This is based on the Feb. 4 comment at the end of #2869. It does not implement author profiles after the exploration.

Screens:

Normal/non-activated:
![Normal/non-activated](https://i.imgur.com/zePR56i.png)

Activated by button press:
![Activated by button press](https://i.imgur.com/lgb5vth.png)

Activated by bar press:
![Activated by bar press](https://i.imgur.com/hjsaBRM.png)

seanlip commented on core/templates/dev/head/pages/exploration_player/exploration_footer_directive.html:
```diff
@@ -162,6 +180,14 @@
+          <li popover-placement="top" popover-template="'popover/feedback'" popover-trigger="click">
+            <a href="" class="btn oppia-navbar-button" tooltip="<['I18N_PLAYER_FEEDBACK_TOOLTIP' | translate]>" tooltip-placement="top">
+              <i class="material-icons">&#xE87F;</i>
+              <span><['I18N_PLAYER_FEEDBACK_TOOLTIP' | translate]></span>
```
Could we go with `span translate="..."` here? (We generally do it like this unless there's no alternative.)

Oishikatta replied:
Fixed, it works fine here.

seanlip commented on core/templates/dev/head/pages/exploration_player/ExplorationFooterDirective.js:
```diff
@@ -40,6 +42,10 @@ oppia.directive('explorationFooter', [function() {
         };
 
+        $scope.windowIsNarrow = windowDimensionsService.isWindowNarrow();
+        console.log($scope.windowIsNarrow);
```
Please remove console.log

Oishikatta replied:
Removed.

seanlip commented on core/templates/dev/head/pages/exploration_player/ExplorationFooterDirective.js:
```diff
@@ -40,6 +42,10 @@ oppia.directive('explorationFooter', [function() {
 
+        $scope.windowIsNarrow = windowDimensionsService.isWindowNarrow();
+        console.log($scope.windowIsNarrow);
+        // add hook
```
?

Oishikatta replied:
I forgot to actually push the second commit.

Oishikatta commented:
The design, overall I have no idea; posted to get comments, it is open to changes.

(a) We can almost certainly do this.

(b) Clicking/tapping on the footer or feedback button is the easiest way to close it, but that may not be intuitive. The main case where screens will be small is landscape mode on phones, and at that point even hiding the footer may not make much difference. I'm not sure.

![](https://i.imgur.com/A6u3UnD.png)

seanlip commented:
Thanks, @Oishikatta!

@jaredsilver @rachelwchen @markhalpin any thoughts on the design aspects of my [previous comment](https://github.com/oppia/oppia/pull/3398#pullrequestreview-36735511) (or anything else?)

seanlip commented:
Also, @Oishikatta, are you planning to file a new issue for author profiles after the exploration? Might be worth doing so that merging this issue closes #2869 and we don't forget to do the other part of it.

rachelwchen commented:
I agree with @seanlip -- let's make the modal more of the focal point of the screen, since that's likely the only point users will be looking at when they type. Maybe we can make the modal even bigger, so that none of the text in the background is showing? (ex: modal will take up ~90% of the screen width)

Oishikatta commented:
I'm not sure on the best way to implement them yet, but I will get at least some mocks posted for the comments so far tomorrow (Thursday).

Oishikatta commented:
My mistake, I meant to write tomorrow (Wednesday).

Oishikatta commented:
Full width and background dimmed (20px space on left and right):

*There isn't a keyboard shown in this one, but on a normal phone it would take up most of the vertical space.
![](https://i.imgur.com/AX0EJAR.png)

I haven't figured out how hiding the bottom bar would work yet.

seanlip commented:
This looks great, IMO. I'm not too bothered about having the footer disappear, actually, so we can probably merge this in as-is. The "files changed" code doesn't yet reflect the mock, right?

Oishikatta commented:
Right, to implement the full width I had to use !important so I'm looking for alternatives to that, and I haven't actually implemented the background dimming yet.

seanlip commented:
OK! I think once you can get it looking like the mock (in a way in which the code is clean), this should be good to go. Let me know if you need advice on the technical design aspects.

shaz13 commented on core/templates/dev/head/pages/exploration_player/exploration_footer_directive.html:
```diff
@@ -23,6 +23,37 @@
+      margin-right: 5px;
+      max-width: 100vw;
+      right: 15px;
+      width: auto !important;
```
> Inline styles are applied by bootstrap-ui javascript, since they have the highest specificity only !important can override them. This can be resolved by updating bootstrap-ui.

Agree, but in here as we are using `max-width`, we can format the width as follows. This code  (without **!important** ) below should work fine. Please confirm. 

```
    color: rgba(0,0,0,0.87);
    margin-left: 5px;
    margin-right: 5px;
    max-width: 97.6vw;
    right: 15px;
```

Oishikatta replied:
That does work, thank you.

anthkris commented:
@Oishikatta Are you awaiting another review?

markhalpin commented:
I think this looks great.

Oishikatta commented:
@anthkris The issue is that I haven't found a way to implement it with good code quality.

Here in FeedbackPopupDirective.js: https://github.com/oppia/oppia/pull/3398/files#diff-1

I used jQuery to add a div for the backdrop when the directive is created, and remove the div when the directive is destroyed.

The div cannot (to my knowledge) be added in the feedback_popup_directive.html due to [CSS stacking contexts](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context), something that was new to me.

The options I can think of:

1) Use a modal instead of a popover

Problems: Top/bottom bars would be covered, may require duplication of code from the feedback popover directive

2) Add a 'backdrop' service and move the dom interaction to it

Problem: Still dom interaction, not much improvement

/cc @seanlip any ideas?

seanlip commented:
I guess there's two things: the popup, and the mask.

For the popup, are you able to reuse the FeedbackPopupDirective but add the necessary styling somehow?

For the mask, perhaps we should have a BackgroundMaskService that controls a boolean property that turns the mask on or off? I think we have a mask when the sidebar is opened on desktop; might be worth looking into how that functions and whether the mask functionality can be pulled out into a separate service.

Oishikatta commented:
For the popup, I am currently doing exactly that: reuse directive and add styling.

For the mask:

I did look for other instances of a background mask, but only found cases where it came from angular-ui's modal directive. I didn't see any from the sidebar.

I could do a BackgroundMaskService with on/off and add a div with ng-if based on the value of that service to the base controller (to put the mask directly under the Body). Does that sound reasonable?

seanlip commented:
Yes, that seems reasonable to me. Thanks!

Oishikatta commented:
I've added the backgroundMaskService to replace the use of jQuery. PR is ready for another look now, thanks.

seanlip commented on core/templates/dev/head/app.js:
```diff
@@ -838,6 +838,20 @@ oppia.factory('windowDimensionsService', ['$window', function($window) {
 
+// Service for enabling a background mask that leaves navigation visible.
+oppia.factory('backgroundMaskService', function() {
+  var maskActive = false;
```
maskIsActive

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/app.js:
```diff
@@ -838,6 +838,20 @@ oppia.factory('windowDimensionsService', ['$window', function($window) {
+  var maskActive = false;
+
+  return {
+    getMaskActive: function() {
```
isMaskActive

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/app.js:
```diff
@@ -838,6 +838,20 @@ oppia.factory('windowDimensionsService', ['$window', function($window) {
+    getMaskActive: function() {
+      return maskActive;
+    },
+    setMaskState: function(state) {
```
activateMask

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/app.js:
```diff
@@ -838,6 +838,20 @@ oppia.factory('windowDimensionsService', ['$window', function($window) {
 }]);
 
+// Service for enabling a background mask that leaves navigation visible.
+oppia.factory('backgroundMaskService', function() {
```
We're moving to [UpperCamelCase for service names](https://docs.google.com/document/d/1ZDmLN66f53WdDPItFChu9Lr37z0dKoqR-ASX8UM5y60/edit), so let's use BackgroundMaskService here.

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/pages/Base.js:
```diff
@@ -31,6 +33,8 @@ oppia.controller('Base', [
     $scope.isSidebarShown = SidebarStatusService.isSidebarShown;
     $scope.closeSidebarOnSwipe = SidebarStatusService.closeSidebar;
 
+    $scope.backgroundMaskActive = backgroundMaskService.getMaskActive;
```
$scope.isBackgroundMaskActive

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/FeedbackPopupDirective.js:
```diff
@@ -126,6 +128,10 @@ oppia.directive('feedbackPopup', [
+
+          $scope.$on('$destroy', function() {
+            backgroundMaskService.setMaskState(false);
+          })
```
add semicolon at end

Oishikatta replied:
Done.

Oishikatta commented:
That was really tricky to track down why the resize hook wasn't working!

It turns out there can only be one window.onresize function, and it's getting defined in another file under exploration_player:

https://github.com/oppia/oppia/blob/74f4a40a5f975b2817251a3a90ff6bbb4e713c77/core/templates/dev/head/pages/exploration_player/ConversationSkinDirective.js#L706

I've changed the windowDimensionsService to use .bind instead of the window.onresize function, so it works as expected now and won't be broken if window.onresize is redefined elsewhere again.

Changes aren't pushed yet, I'll comment when it's ready.

Oishikatta commented:
I think the whitespace was related to a [bug in angular-ui-bootstrap 0.13.4, fixed in 0.14.0](https://github.com/angular-ui/bootstrap/commit/81a38d72613d82bb4d8b5ab44e3bde704a59b056#diff-e8665eb306ddd26399e5b80f00dae5bbL153). The same issue is present with top feedback button in develop. I was able to prevent it by changing the feedback popup to use ng-if instead of ng-show though (will apply to top button as well).

I fixed the other things, the last thing I'm unsure on is that between 651px ~ 768px, both the top and bottom feedback buttons are visible. Do you think we should make the bottom one only visible at 650 and under?

Thanks.

seanlip commented:
Yeah, I think only one feedback button should show at any given screen width. Would it be possible to change that accordingly? I do not really have a view on what the exact cutoff should be, but 650 seems reasonable to me.

Also, well done on figuring out the resize issue! That's pretty subtle...

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -588,6 +588,19 @@ textarea {
 }
 
 /*
+  Styles for backgroundMaskService.
```
nit: b --> B

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/FeedbackPopupDirective.js:
```diff
@@ -126,6 +132,10 @@ oppia.directive('feedbackPopup', [
           };
+
+          $scope.$on('$destroy', function() {
+            BackgroundMaskService.activateMask(false);
```
This reads a bit weirdly. Maybe have two methods: activateMask and deactivateMask.

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/pages/exploration_player/exploration_footer_directive.html:
```diff
@@ -16,17 +16,42 @@
   }
 
+  .oppia-exploration-footer .mobile-feedback {
+    margin-bottom: 0px;
```
If the value is 0, let's omit the units.

Oishikatta replied:
Done.

seanlip commented:
FYI, a feedback-related e2e test seems to be failing.

seanlip commented:
Hi @Oishikatta -- looks like the ng-show thing fixes the e2e test. Perhaps the best long-term fix is to upgrade the ui-bootstrap lib. Shall we file that as a separate issue, and get this in in the meantime?

Oishikatta commented:
I wanted to figure out why the test wasn't working with ng-if, but wasn't able to in the time I spent on it. I'll file a separate issue for upgrading ui-bootstrap.

I've added the code to switch at 650px and to address the newer comments.

seanlip commented on core/templates/dev/head/pages/exploration_player/FeedbackPopupDirective.js:
```diff
@@ -47,6 +49,10 @@ oppia.directive('feedbackPopup', [
           $scope.feedbackPopoverId = (
             'feedbackPopover' + Math.random().toString(36).slice(2));
 
+          if ( windowDimensionsService.getWidth() < 651 ) {
```
Nit: remove spaces after '(' and before ')'

Can we make 651 a named constant defined in app.js? It appears in several places and we should make it such that it appears just once in the codebase. Alternatively we could write a custom function in windowDimensionsService to check narrowness for this purpose.

Oishikatta replied:
The 651px comes from here in CSS: https://github.com/Oishikatta/oppia/blob/271075d67a6cac7a8cbe5ffbd6a99c00957fc95f/core/templates/dev/head/css/oppia.css#L907

I don't think we have a way to define constants for there though. As far as I can tell it is only visible in the exploration player, even though it style appears in the exploration player, exploration editor, collection player, collection editor, and admin page.

So I added the function and named it isExplorationPlayerNavHidden, also removed the extra spaces.

seanlip commented:
LGTM. Thanks @Oishikatta! If you're up for it, I think one thing that may be worth doing is to add a "NOTE TO DEVELOPERS" in the places where 651 is referenced, asking to make sure the other places are also updated when that value is updated. But otherwise this looks great.

Oishikatta commented:
Sure, I'll add the note in a few minutes.
